### PR TITLE
BI-10168 refactor to separate mapping concerns

### PR DESCRIPTION
**Pull request by phudson-ch**
- Make OrderMessageHandler an idempotent receiver
- Move test synchronisation logic into Spring aspects
- Separate order message consumer into separate class
- Eliminate use of inheritance in order kind mapper implementations
- Added OrdersApiDetails facade [interface]
- Added OrdersApiDetailsBuilder to adapt OrdersApi property access
- Refactored application classes to use OrdersApiDetails i.e. in place of OrdersApi

[BI-10168 Order notification sender should not retry messages where orders API returns HTTP 404](https://companieshouse.atlassian.net/browse/BI-10168)

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/ordernotification/emailsendmodel/OrdersApiDetailsCommonFieldsMapper.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+class OrdersApiDetailsCommonFieldsMapper {
+    private final EmailConfiguration emailConfiguration;
+
+    public OrdersApiDetailsCommonFieldsMapper(EmailConfiguration emailConfiguration) {
+        this.emailConfiguration = emailConfiguration;
+    }
+
+    void mapCommonFields(OrderModel orderModel, OrdersApiDetails ordersApiDetails) {
+        orderModel.setTo(ordersApiDetails.getOrderEmail());
+        orderModel.setSubject(MessageFormat.format(emailConfiguration.getConfirmationMessage(), ordersApiDetails.getOrderReference()));
```
Comment (phudson-ch) [inline src/main/java/uk/gov/companieshouse/ordernotification/emailsendmodel/OrdersApiDetailsCommonFieldsMapper.java:19]: Check whether orderReference can be on OrderApiDetails?

File: src/test/java/uk/gov/companieshouse/ordernotification/emailsendmodel/CertificateOrderNotificationMapperTest.java [hunk 2]
```diff
@@ -1,64 +1,70 @@
...
+
+    @BeforeEach
+    void setup() {
+        certificateOrderNotificationMapper = new CertificateOrderNotificationMapper(config,
+                mapperFactory);
     }
 
     @Test
-    void testCertificateOrderNotificationMapperReturnsMessageType() {
+    void testCorrectlyMapsCertificateOrderApiMapsToOrderDetails() {
```
Comment (phudson-ch) [inline src/test/java/uk/gov/companieshouse/ordernotification/emailsendmodel/CertificateOrderNotificationMapperTest.java]: Provide more descriptive test name use DisplayName annotation

File: src/test/java/uk/gov/companieshouse/ordernotification/emailsendmodel/CertificateOrderNotificationMapperTest.java [hunk 3]
```diff
@@ -1,64 +1,70 @@
...
         when(config.getCertificate()).thenReturn(emailDataConfig);
+        when(emailDataConfig.getMessageId()).thenReturn(TestConstants.MESSAGE_ID);
         when(emailDataConfig.getMessageType()).thenReturn(TestConstants.MESSAGE_TYPE);
+        when(certificateItemOptionsApi.getCompanyType()).thenReturn("ltd");
+        when(mapperFactory.getCertificateOptionsMapper("ltd")).thenReturn(certificateOptionsMapper);
+        when(certificateOptionsMapper.generateEmailData(ordersApiDetails)).thenReturn(orderModel);
 
         //when
-        String actual = certificateOrderNotificationMapper.getMessageType();
+        uk.gov.companieshouse.ordernotification.emailsendmodel.OrderDetails orderDetails = certificateOrderNotificationMapper.map(
```
Comment (phudson-ch) [inline src/test/java/uk/gov/companieshouse/ordernotification/emailsendmodel/CertificateOrderNotificationMapperTest.java]: Remove fully qualified path

File: src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrderMessageHandlerTest.java [hunk 4]
```diff
@@ -40,6 +43,7 @@ void testHandleMessagePublishesEvent() {
         // given
         SendOrderNotificationEvent event = new SendOrderNotificationEvent("ORD-12345-678", 1);
         when(eventFactory.createEvent(message)).thenReturn(event);
+        when(messageFilter.include(message)).thenReturn(true);
```
Comment (phudson-ch) [inline src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrderMessageHandlerTest.java:47]: Add negative case

File: src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerIntegrationTest.java [hunk 5]
```diff
@@ -29,87 +39,89 @@
...
-        container.stop();
-    }
-
     @Test
     void testHandlesCertificateOrderReceivedMessage() throws ExecutionException, InterruptedException, IOException {
         //given
         client.when(request()
-                .withPath(TestConstants.ORDER_NOTIFICATION_REFERENCE)
-                .withMethod(HttpMethod.GET.toString()))
+                        .withPath(TestConstants.ORDER_NOTIFICATION_REFERENCE)
```
Comment (phudson-ch) [inline src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerIntegrationTest.java]: Check source for tabs

File: src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerIntegrationTest.java [hunk 6]
```diff
@@ -29,87 +39,89 @@
...
+        TestEnvironmentSetupHelper.setEnvironmentVariable("PAYMENTS_API_URL",
+                "http://" + container.getHost() + ":" + container.getServerPort());
+    }
+
+    @AfterAll
+    static void after() {
+        container.stop();
+    }
+
+    private static OrderReceived getOrderReceived() {
```
Comment (phudson-ch) [inline src/test/java/uk/gov/companieshouse/ordernotification/ordersconsumer/OrdersKafkaConsumerIntegrationTest.java]: Move to end

**Review by kinpang-CH — APPROVED** [review]

**Review by dlgroves-ch — APPROVED** [review]

**Pull request merged**